contracts/core/modules/ExchangeRedeemModule.sol

```diff
@@ -0,0 +1,193 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+*/
+
+pragma solidity 0.5.4;
+pragma experimental "ABIEncoderV2";
+
+import { ReentrancyGuard } from "openzeppelin-solidity/contracts/utils/ReentrancyGuard.sol";
+import { SafeMath } from "openzeppelin-solidity/contracts/math/SafeMath.sol";
+
+import { ExchangeExecution } from "./lib/ExchangeExecution.sol";
+import { ExchangeIssueLibrary } from "../lib/ExchangeIssueLibrary.sol";
+import { ExchangeHeaderLibrary } from "../lib/ExchangeHeaderLibrary.sol";
+import { ExchangeValidationLibrary } from "../lib/ExchangeValidationLibrary.sol";
+import { ExchangeWrapperLibrary } from "../lib/ExchangeWrapperLibrary.sol";
+import { IExchangeWrapper } from "../interfaces/IExchangeWrapper.sol";
+import { ISetToken } from "../interfaces/ISetToken.sol";
+import { LibBytes } from "../../external/0x/LibBytes.sol";
+import { ModuleCoreState } from "./lib/ModuleCoreState.sol";
+
+
+/**
+ * @title ExchangeRedeemModule
+ * @author Set Protocol
+ *
+ * The Exchange Redeem Module facilitates the exchangeRedeem function which allows
+ * the redemption of a Set using exchange orders into specified received tokens
+ */
+contract ExchangeRedeemModule is
+    ModuleCoreState,
+    ExchangeExecution,
+    ReentrancyGuard
+{
+    using SafeMath for uint256;
+
+    /* ============ Events ============ */
+
+    event LogExchangeRedeem(
+        address setAddress,
+        address indexed callerAddress,
+        uint256 quantity,
+        address[] receiveTokens,
+        uint256[] receiveTokenAmounts
+    );
+
+    /* ============ Constructor ============ */
+
+    /**
+     * Constructor function for ExchangeRedeemModule
+     *
+     * @param _core                The address of Core
+     * @param _vault               The address of Vault
+     */
+    constructor(
+        address _core,
+        address _vault
+    )
+        public
+        ModuleCoreState(
+            _core,
+            _vault
+        )
+    {}
+
+    /* ============ Public Functions ============ */
+
+    /**
+     * Redeems a Set and performs trades via exchange wrappers to acquire receive tokens to the caller
+     *
+     * @param _exchangeInteractData                A Struct containing exchange interact metadata
+     * @param _orderData                           Bytes array containing the exchange orders to execute
+     */
+    function exchangeRedeem(
+        ExchangeIssueLibrary.ExchangeIssueParams memory _exchangeInteractData,
+        bytes memory _orderData
+    )
+        public
+        nonReentrant
+    {
+        // Validate ExchangeRedeemParams
+        validateExchangeIssueParams(_exchangeInteractData);
```

Comment:
Can we give this a more general name? The juxtapostion between the comment and this is kinda confusing

Reply:
Yeah in a separate PR, I'm going to rename everything to `exchangeIssuance` (requires set-protocol-utils updates)

Reply:
I ended up adding it to this PR